### ISSUE TITLE
where is the training model of PMC? how to use it?

### ISSUE THREAD

**Issue by XinLang2019** [Tencent-RoboticsX/lifelike-agility-and-play#7]
i don't know where the training model of PMC is. i want to use GO2 robot. So can you tell me how to use our training model?

**Comment by slivermoda** [comment]
The trained PMC model is located at data/models/primitive_level.model

This is the trained PMC model of the MAX robot in the paper. If you use Go2, you need to import Go2's urdf and follow the PMC training procedure to train for your specific robot.

**Comment by XinLang2019** [comment]
> The trained PMC model is located at data/models/primitive_level.model
> 
> This is the trained PMC model of the MAX robot in the paper. If you use Go2, you need to import Go2's urdf and follow the PMC training procedure to train for your specific robot.

Thank you for your replay! only one question, i want to know how much the figure of reward_sum represents this pmc model can be used? Now i try to retrain a PMC model  by using your max robot (the reward_sum is about 23), but the performance of the model is bad (can't move like your trained PMC model).

**Comment by slivermoda** [comment]
> > The trained PMC model is located at data/models/primitive_level.model
> > This is the trained PMC model of the MAX robot in the paper. If you use Go2, you need to import Go2's urdf and follow the PMC training procedure to train for your specific robot.
> 
> Thank you for your replay! only one question, i want to know how much the figure of reward_sum represents this pmc model can be used? Now i try to retrain a PMC model by using your max robot (the reward_sum is about 23), but the performance of the model is bad (can't move like your trained PMC model).

Using exact the same reward setting in the PMC codes, our trained model gets 120+ reward_sum

**Comment by XinLang2019** [comment]
![2024-09-03 10-08-46 的屏幕截图](https://github.com/user-attachments/assets/5d143117-1492-49d7-8f83-25650a947d55)
this graph shows the figure of reward_sum, it seems to have levelled off at around 22, which is a long way from 120 you mentioned. not sure why(the training parameters have not been modified), is the parameters have been changed in your repository.

**Comment by slivermoda** [comment]
example_pmc_train.sh shows an example training script, and it keeps almost the same hyper parameters that we used in our experiments, except for those w.r.t. computation scales. Specifically, 

rm_size, batch_size and number of parallel running actors 

are larger in our experiments, and the choice of these parameters depend on your computing resources. For tensorboard logs, you might need to focus on the consuming fps and receiving fps that are important metrics showing how your training scales. Normally, 

consuming fps >= receiving fps >= 1 / 10 * consuming fps

is a reasonable config of the training.

**Comment by XinLang2019** [comment]
> example_pmc_train.sh shows an example training script, and it keeps almost the same hyper parameters that we used in our experiments, except for those w.r.t. computation scales. Specifically,
> 
> rm_size, batch_size and number of parallel running actors
> 
> are larger in our experiments, and the choice of these parameters depend on your computing resources. For tensorboard logs, you might need to focus on the consuming fps and receiving fps that are important metrics showing how your training scales. Normally,
> 
> consuming fps >= receiving fps >= 1 / 10 * consuming fps
> 
> is a reasonable config of the training.

what's the name of number of parallel actors in your code? is 'batch_size' in 'actor_policy_config'?

**Comment by slivermoda** [comment]
> > example_pmc_train.sh shows an example training script, and it keeps almost the same hyper parameters that we used in our experiments, except for those w.r.t. computation scales. Specifically,
> > rm_size, batch_size and number of parallel running actors
> > are larger in our experiments, and the choice of these parameters depend on your computing resources. For tensorboard logs, you might need to focus on the consuming fps and receiving fps that are important metrics showing how your training scales. Normally,
> > consuming fps >= receiving fps >= 1 / 10 * consuming fps
> > is a reasonable config of the training.
> 
> what's the name of number of parallel actors in your code? is 'batch_size' in 'actor_policy_config'?

There is no variable controling the number of parallel actors in the codes. You need to launch actors separately in parallel threads or processes.

In our case, we used 4 GPUs for the learner and each GPU consumes samples with around 2.5e+5 fps. To match the scale 

consuming fps >= receiving fps >= 1 / 10 * consuming fps

as mentioned above, we launched 1000 actors (one CPU core for each actor) for each GPU with k8s on Cloud, to reach a receiving fps around 0.5e+5.

**Comment by XinLang2019** [comment]
> > > example_pmc_train.sh shows an example training script, and it keeps almost the same hyper parameters that we used in our experiments, except for those w.r.t. computation scales. Specifically,
> > > rm_size, batch_size and number of parallel running actors
> > > are larger in our experiments, and the choice of these parameters depend on your computing resources. For tensorboard logs, you might need to focus on the consuming fps and receiving fps that are important metrics showing how your training scales. Normally,
> > > consuming fps >= receiving fps >= 1 / 10 * consuming fps
> > > is a reasonable config of the training.
> > 
> > 
> > what's the name of number of parallel actors in your code? is 'batch_size' in 'actor_policy_config'?
> 
> There is no variable controling the number of parallel actors in the codes. You need to launch actors separately in parallels threads or processes.
> 
> In our case, we used 4 GPUs for the learner and each GPU consumes samples with around 2.5e+5 fps. To match the scale
> 
> consuming fps >= receiving fps >= 1 / 10 * consuming fps
> 
> as mentioned above, we launched 1000 actors (one CPU core for each actor) for each GPU with k8s on Cloud, to reach a receiving fps around 0.5e+5.

So, you mean that we should launch other learner in different terminals (or change the parameter "batch_worker_num=4"), and how launch 1000 actor?(which parapeter should be changed?)

**Comment by slivermoda** [comment]
> > > > example_pmc_train.sh shows an example training script, and it keeps almost the same hyper parameters that we used in our experiments, except for those w.r.t. computation scales. Specifically,
> > > > rm_size, batch_size and number of parallel running actors
> > > > are larger in our experiments, and the choice of these parameters depend on your computing resources. For tensorboard logs, you might need to focus on the consuming fps and receiving fps that are important metrics showing how your training scales. Normally,
> > > > consuming fps >= receiving fps >= 1 / 10 * consuming fps
> > > > is a reasonable config of the training.
> > > 
> > > 
> > > what's the name of number of parallel actors in your code? is 'batch_size' in 'actor_policy_config'?
> > 
> > 
> > There is no variable controling the number of parallel actors in the codes. You need to launch actors separately in parallels threads or processes.
> > In our case, we used 4 GPUs for the learner and each GPU consumes samples with around 2.5e+5 fps. To match the scale
> > consuming fps >= receiving fps >= 1 / 10 * consuming fps
> > as mentioned above, we launched 1000 actors (one CPU core for each actor) for each GPU with k8s on Cloud, to reach a receiving fps around 0.5e+5.
> 
> So, you mean that we should launch other learner in different terminals (or change the parameter "batch_worker_num=4"), and how launch 1000 actor?(which parapeter should be changed?)

No, you do not need to run multiple learners, and you do not need to change 'batch_worker_num' which is for prefetch logic, neither. In TLeague, you only need to launch one learner while it can equip with multiple GPUs, then for each GPU, you can launch multiple actors by opening many terminals or launching many pods in k8s cluster on cloud. The parameter for an actor you need to pay attention to is the --learner_addr, which binds it to a specific GPU (via different ports).

I would recommend to read the TLeague paper and its usage, for better understanding the answers for these questions.

**Comment by XinLang2019** [comment]
> > > > > example_pmc_train.sh shows an example training script, and it keeps almost the same hyper parameters that we used in our experiments, except for those w.r.t. computation scales. Specifically,
> > > > > rm_size, batch_size and number of parallel running actors
> > > > > are larger in our experiments, and the choice of these parameters depend on your computing resources. For tensorboard logs, you might need to focus on the consuming fps and receiving fps that are important metrics showing how your training scales. Normally,
> > > > > consuming fps >= receiving fps >= 1 / 10 * consuming fps
> > > > > is a reasonable config of the training.
> > > > 
> > > > 
> > > > what's the name of number of parallel actors in your code? is 'batch_size' in 'actor_policy_config'?
> > > 
> > > 
> > > There is no variable controling the number of parallel actors in the codes. You need to launch actors separately in parallels threads or processes.
> > > In our case, we used 4 GPUs for the learner and each GPU consumes samples with around 2.5e+5 fps. To match the scale
> > > consuming fps >= receiving fps >= 1 / 10 * consuming fps
> > > as mentioned above, we launched 1000 actors (one CPU core for each actor) for each GPU with k8s on Cloud, to reach a receiving fps around 0.5e+5.
> > 
> > 
> > So, you mean that we should launch other learner in different terminals (or change the parameter "batch_worker_num=4"), and how launch 1000 actor?(which parapeter should be changed?)
> 
> No, you do not need to run multiple learners, and you do not need to change 'batch_worker_num' which is for prefetch logic, neither. In TLeague, you only need to launch one learner while it can equip with multiple GPUs, then for each GPU, you can launch multiple actors by opening many terminals or launching many pods in k8s cluster on cloud. The parameter for an actor you need to pay attention to is the --learner_addr, which binds it to a specific GPU (via different ports).
> 
> I would recommend to read the TLeague paper and its usage, for better understanding the answers for these questions.

think you for your reply!

**Comment by CoderWangcai** [comment]
@XinLang2019 May I ask if you have completed the PMC phase training based on Go2? Could I also ask for the code for converting mocap data to Go2, as well as the code modifications needed for PMC training? Thank you.
